core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java

```diff
@@ -224,12 +159,39 @@ protected boolean shardAvailable(ShardId shardId) {
         return shardAvailable(getShard(shardId));
     }
 
+    /** returns how much heap this shard is using for its indexing buffer */
+    protected long getIndexBufferRAMBytesUsed(ShardId shardId) {
+        IndexShard shard = getShard(shardId);
+        if (shard == null) {
+            return 0;
+        }
+
+        return shard.getIndexBufferRAMBytesUsed();
+    }
+
+    /** ask this shard to refresh, in the background, to free up heap */
+    protected void refreshShardAsync(ShardId shardId) {
+        IndexShard shard = getShard(shardId);
+        if (shard != null) {
+            shard.refreshAsync("memory");
```

Comment:
I wonder if we should keep IndexShard clean just call refresh from a thread here?